File: projects/therobotknows.com/app/frontend/src/lib/api/graph.ts
import { isMockMode } from "./config";
import { request } from "./client";
import { entriesApi } from "./entries";
import type { EntryType, EntryStatus } from "@/types/api";

export interface GraphNode {
  id: string;
  slug?: string | null;
  label: string;
  type: EntryType;
  status: EntryStatus;
  era?: string | null;
  region?: string | null;
}

export interface GraphEdge {
  id?: string;
  source: string;
  target: string;
  relationship: string;
  excerpt?: string | null;
}

export interface GraphParams {
  type?: string;
  status?: string;
  tag?: string;
  era?: string;
  region?: string;
}

export const graphApi = {
  async get(
    universeId: string,
    params: GraphParams = {},
  ): Promise<{ nodes: GraphNode[]; edges: GraphEdge[] }> {
    if (isMockMode()) {
      const { entries } = await entriesApi.list(universeId, { per_page: 100 });
      const filtered = entries.filter((e) => {
        if (params.type && e.type !== params.type) return false;
        if (params.status && e.status !== params.status) return false;
        return true;
      });
      const nodes: GraphNode[] = filtered.map((e) => ({
        id: e.id,
        slug: e.slug,
        label: e.title,
        type: e.type,
        status: e.status,
        era: e.era,
        region: e.region,
      }));
      const edges: GraphEdge[] = [];
      for (const e of filtered) {
        for (const l of e.links || []) {
          edges.push({
            id: l.id,
            source: l.source_entry_id,
            target: l.target_entry_id,
            relationship: l.relationship,
            excerpt: l.excerpt,
          });
        }
      }
      return { nodes, edges };
    }
    const q = new URLSearchParams();
    Object.entries(params).forEach(([k, v]) => {
      if (v) q.set(k, v);
    });
    const qs = q.toString();
    return request(
      `/api/v1/universes/${encodeURIComponent(universeId)}/graph${qs ? `?${qs}` : ""}`,
    );
  },
};

File: projects/therobotknows.com/app/backend/lib/therobotknows/graph.ex
defmodule Therobotknows.Graph do
  @moduledoc "Knowledge graph payload from entries + entry_links."

  import Ecto.Query
  alias Therobotknows.Repo
  alias Therobotknows.Universes
  alias Therobotknows.Schema.Canon.{Entry, EntryLink}

  def graph(universe_id_or_slug, user_id, opts \\ []) do
    with {:ok, universe, _role} <- Universes.authorize(universe_id_or_slug, user_id, :viewer) do
      type = Keyword.get(opts, :type)
      status = Keyword.get(opts, :status)
      tag = Keyword.get(opts, :tag)
      era = Keyword.get(opts, :era)
      region = Keyword.get(opts, :region)

      entries_q =
        from e in Entry,
          where: e.universe_id == ^universe.id and is_nil(e.deleted_at)

      entries_q =
        entries_q
        |> maybe_eq(:type, type)
        |> maybe_eq(:status, status)
        |> maybe_eq(:era, era)
        |> maybe_eq(:region, region)

      entries =
        if tag && tag != "" do
          from(e in entries_q,
            join: et in Therobotknows.Schema.Canon.EntryTag,
            on: et.entry_id == e.id,
            join: t in Therobotknows.Schema.Canon.Tag,
            on: t.id == et.tag_id,
            where: t.slug == ^tag
          )
          |> Repo.all()
        else
          Repo.all(entries_q)
        end

      entry_ids = Enum.map(entries, & &1.id)

      links =
        from(l in EntryLink,
          where:
            l.universe_id == ^universe.id and l.source_entry_id in ^entry_ids and
              l.target_entry_id in ^entry_ids
        )
        |> Repo.all()

      nodes =
        Enum.map(entries, fn e ->
          %{
            id: e.id,
            slug: e.slug,
            label: e.title,
            type: e.type,
            status: e.status,
            era: e.era,
            region: e.region
          }
        end)

      edges =
        Enum.map(links, fn l ->
          %{
            id: l.id,
            source: l.source_entry_id,
            target: l.target_entry_id,
            relationship: l.relationship,
            excerpt: l.excerpt
          }
        end)

      {:ok, %{nodes: nodes, edges: edges}}
    end
  end

  defp maybe_eq(q, _f, nil), do: q
  defp maybe_eq(q, _f, ""), do: q
  defp maybe_eq(q, f, v), do: from(e in q, where: field(e, ^f) == ^v)
end
